### PR TITLE
🔀 Implemented RTK Query API

### PR DESCRIPTION
**Pull request by shon-button**
[Addresses](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/119)

✨ Components:
1. Add Redux slice for API instance: redux\features\api
2. Add Redux slice for authentication state: redux\features\auth
3. Add Redux slice for user state: redux\features\user
4. Add client side component: app\components\auth\Login
5. Add server side component: app\components\user\Profile
6. Add route for mock login: app\api\auth\route.ts
7. Add route for mock users: app\api\user\route.ts

 🚀 Impact:
   - Configured RTK Query actions, reducers, and selectors for interacting with the API
   - Added mock authentication-related state management to the Redux store
   - Added mock user-related state management to the Redux store
   - Created a client-side component to use RTK Query to fetch mock authentication data including features of RTK Query loading and error state
   - Created a server-side component to use RTK Query to fetch mock user data including features of RTK Query caching and loading

 🗒️ Notes:
   - In SSR, when you use RTK Query, direct access to the Redux store is necessary  to fetch data before rendering the page. 
   - If you access the Redux store directly in SSR without taking precautions, there's a potential risk of data leakage between different requests. For example, data from one user's request might temporarily be accessible to another user's request, leading to a security issue.
   - To mitigate the risk of data leakage, it's recommended to "purge" or clear the global Redux store at the root layout component of your application before rendering a new page for a different user. This ensures that each request starts with a clean slate and doesn't accidentally share data with other requests.

**Demo steps**:

1. Open terminal in cas-registration\client folder
2. Start the development server:
   ```
   yarn dev
   ```
3. In your web browser, navigate to `http://localhost:3000/`

4.1 Interact with the Login Page:   
   - To error: click "Login"
   Error message should display based on RTK Query error status
image

4.2 Interact with the Login Page:
   - To Login: Enter text in email and password fields then click "Login"
Loading message should display based on RTK Query loading status
image

User profile should display based on RTK Query success status

5. User profile is a SSR page that displays auth token based on store cache value, and calls a RTK Query from server side
image

### CODE REVIEW

File: client/redux/features/api/slice.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    //                       ResultType    QueryArg
+    //                           v           v
+    login: builder.mutation<LoginResponse, LoginRequest>({
+      query: (credentials) => ({
+        url: "login",
+        method: "POST",
+        body: credentials,
+      }),
+    }),
+    // Define a users query
```
Comment (BCerki) [inline client/redux/features/api/slice.ts:44]: We should do all our CRUD gets etc. from this api slice?

File: client/redux/features/user/selectors.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { apiSlice } from "@/redux/features/api/slice";
+
+// 🎯 Selector function, allowing us to retrieve a value from the Redux state.
+
+// Calling `someEndpoint.select(someArg)` generates a new selector that will return the query result object for a query with those parameters
+export const selectUserData = apiSlice.endpoints.getUsers.select(null);
```
Comment (BCerki) [inline client/redux/features/user/selectors.ts:6]: How is this different from the getUsers hook?

File: client/redux/features/user/slice.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+// 🍕 Slice: RTK’s createSlice to create a Redux slice
+export const userSlice = createSlice({
+  name: "user",
+  initialState,
+  reducers: {
+    resetUserData: () => {
+      return initialState;
+    },
+  },
```
Comment (BCerki) [inline client/redux/features/user/slice.ts:17]: In the docs I sometimes see the getUsers, getUsersById etc. here. Now that we have RTK, are the reducers here only going to be non-CRUD-related things?

**Review by BCerki — COMMENTED** [review]
I have some questions about the redux setup; if a walkthrough is easier than answering here, just let me know!

File: client/redux/features/api/slice.ts (hunk 1 above)
Comment (shon-button) [reply client/redux/features/api/slice.ts:44]: Typically, one API slice **per base URL** to which the application communicates.
In this demo, I made a generically name apiSlice inferring only 1 base URL but, I suspect real use cases will require multiple slices for example authAPI, for crudAPI , for reportAPI etc.  each having a different base URL and appropriate namespace.

File: client/redux/features/user/slice.ts (hunk 3 above)
Comment (shon-button) [reply client/redux/features/user/slice.ts:17]: All CRUD actions are possible with RTK Query endpoints.
Query type endpoints fetch data from the server and mutation type endpoints send data to the server.  A mutation hook, such as 'useLoginMutation', sends data to the "server", in this code base example it sends it to app\api\auth\route.ts, and this payload data can then be used for create, update, or delete operations.

File: client/redux/features/user/selectors.ts (hunk 2 above)
Comment (shon-button) [reply client/redux/features/user/selectors.ts:6]: The createApi endpoint definition getUsers, exposed as useGetUsersQuery, is responsible for defining the query and managing the data fetching process, while the selector selectUserData is responsible for selecting and providing access to the fetched data within components. 
I have been remiss not to have included an example of using the selector after the hook dispatch.

**Review by andrea-williams — APPROVED** [review]

File: client/app/api/login/route.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { NextResponse } from "next/server";
+import { generateRandomToken, parseEmail } from "@/utils/helpers";
```
Comment (Sepehr-Sobhani) [inline client/app/api/login/route.ts:2]: nit: unused `generateRandomToken`

File: client/app/api/login/route.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+export async function POST(req: Request, res: Response) {
```
Comment (Sepehr-Sobhani) [inline client/app/api/login/route.ts:4]: nit: unused `res`

File: client/app/components/auth/Login.tsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        if (error && typeof error === "object") {
+          // Narrow the type to FetchBaseQueryError
+          if ("status" in error) {
+            // You can now access properties of FetchBaseQueryError
+            const errMsg = JSON.stringify(error.data);
+            console.error("RTK Query Error ❌:", error.status, errMsg);
+          }
+        }
+      }
+    } catch (error) {
```
Comment (Sepehr-Sobhani) [inline client/app/components/auth/Login.tsx:57]: nit: 'error' is already declared in the upper scope

File: client/app/api/login/route.ts (hunk 4 above)
Comment (shon-button) [reply client/app/api/login/route.ts]: 👍 nits accepted :)

got to check my ESLint!

**Review by Sepehr-Sobhani — APPROVED** [review]
This looks just great 🤩 This is a fantastic tutorial on Next.js and Redux for me 🥇. Although I had some minor comments, they do not diminish the value of this PR.

**Pull request closed without merging**